feat(outcomes): new orgstats endpoints, outcomes snuba module

JoshFerge opened the pull request:
- API description, architecture: https://www.notion.so/sentry/Sentry-outcomes-consumer-module-backend-OrgStats-changes-ee78192ee4f443be9be66009968de06f
- Adds an endpoint and snuba module to do queries against our outcomes data in clickhouse.
- The architecture/querying is very similar to the sessions_v2 endpoint, and shares code and extends where applicable.
- It differs primarily in how you filter: instead of using the `query` field, filters for `outcome`, `category`, and `reason` are passed directly as query params. This seemed more sensible than trying to extend our querying grammar to work with outcomes
    - it looks like sessions has some auto-aggregation and dataset switching on the snuba side, which outcomes does not have, so had to add a small amount of logic to handle this query side.
- I typed the `outcomes.py` file and added it to the `mypy.ini` file, would love thoughts on how I typed it. also added minimal typing elsewhere to have type compatible imports.


Things to test in staging:

- [x] performance of queries in large orgs, project wise and event wise. 
- [x] performance of queries for minute rollups
- [x] performance of queries with different filters applied
- [ ] test issue where you can't see all projects on stats page

### TODO
- [ ] mark old outcomes TSDB module as deprecated through code comment or otherwise
- [x] final small TODOs in code

JoshFerge committed:
intiail prototyping snuba module

JoshFerge committed:
more prototyping

JoshFerge committed:
Merge branch 'master' into joshferge/outcomes-snuba-module

JoshFerge committed:
zerofill working

JoshFerge committed:
return correct format from endpoint

JoshFerge committed:
make output like billingstats

JoshFerge committed:
get working with new schema

JoshFerge committed:
write initial tests

JoshFerge committed:
initial tests working

JoshFerge committed:
add todos

JoshFerge committed:
merge

JoshFerge committed:
dropped reason functionality working

JoshFerge committed:
cleanup code

JoshFerge committed:
rename src var

JoshFerge committed:
rename funcs, add outcome classmethod

JoshFerge committed:
code cleanups

JoshFerge committed:
add todo for ts

JoshFerge committed:
revert old project stats endpoint

JoshFerge committed:
code cleanup

JoshFerge commented on src/sentry/api/urls.py:
```diff
@@ -1210,6 +1214,21 @@
                     OrganizationStatsEndpoint.as_view(),
                     name="sentry-api-0-organization-stats",
                 ),
+                url(
```
need to decide on how we do versioning with this API. we intend to keep the old one around in case some customers are using it to generate reports. for now, `_v2` is added there.

dbennett-sentry replied:
Hi! Long time listener, first time caller. Can we check the logs to confirm anyone outside Sentry is directly calling the original API?

JoshFerge replied:
looks like 11 orgs have called this API in the last 90 days in sentry SaaS using the best proxy i could for non-sentry-webpage calls.

JoshFerge committed:
Merge branch 'master' into joshferge/outcomes-snuba-module

JoshFerge committed:
fix deepcopy, add todos

JoshFerge committed:
remove comment

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,225 @@
+        response = {
+            "statsErrors": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsTransactions": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
+            "statsAttachments": defaultdict(lambda: deepcopy(DEFAULT_TS_VAL)),
```
Nit: instead of the constant `DEFAULT_TS_VAL`, defining

```python
def create_default_ts_value():
    return {...}
```

would let you just write `defaultdict(create_default_ts_value)`, avoiding the use of both `lambda` and `deepcopy`.

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,225 @@
+    # https://stackoverflow.com/a/3233356
+    for k, v in u.items():
+        if isinstance(v, collections.abc.Mapping):
+            d[k] = nested_update(d.get(k, {}), v)
```
Recursing to arbitrary depth seems squirrelly to me. In the places where this is called, would it be possible to unpack the appropriate number of levels instead?

It looks like this is meant to explore a structure like
```python
{
    "accepted": {"quantity": 0, "times_seen": 0},
    "filtered": {"quantity": 0, "times_seen": 0},
    "dropped": {
        "overQuota": {"quantity": 0, "times_seen": 0},
        "spikeProtection": {"quantity": 0, "times_seen": 0},
        "other": {"quantity": 0, "times_seen": 0},
    },
}
```
and hit a `{"quantity": 0, "times_seen": 0}` object whether it's one or two layers deep. See my other comment about introducing classes regarding that.

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,225 @@
+        for row in result:
+            nested_update(
+                response[CATEGORY_NAME_MAP[row["category"]]][row["time"]], map_row_to_format(row)
+            )
```
I'm seeing a few readability issues here, which I think could be improved in part by introducing one or more classes to model the data that you're building to put into the response, and only converting it to a raw serializable value at the end.

If I'm getting this correctly, there are three layers of structure to model. The topmost is the data categories, the second is the structure of `DEFAULT_TS_VAL`, and the third is `{"quantity": 0, "times_seen": 0}`. If I were writing this from scratch and trying to maximize clarity, I'd probably write one class for each of those three.

One advantage is that the second class could just be a flat structure of five values ("accepted", "filtered", "overQuota", "spikeProtection", "other"), where you only worry about wrapping the last three into the `dropped` dict at the final serialization step.

If you're interested in undertaking this as a refactoring, I'll pair-program on it if you like.

JoshFerge replied:
yes, good call on all of this. will give modeling this w/ classes a shot and making it more readable, will reach out for a pair-program if i'm feeling anything is ambiguous. thank you!

JoshFerge committed:
OO readability refactor

JoshFerge committed:
add todos

markstory commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+
+
+# TODO: good names for these classes
+class StatsResponse:
```
Other parts of the application have this kind of logic in `sentry.api.serializers`. There are also some useful base classes in `sentry.api.serializers.snuba` that help with zerofilling timeseries results.

JoshFerge replied:
would it be okay for these classes to live in `serializers/models` or do we limit those to just ones with corresponding django serializers? i see in a few other places we create a top level folder for endpoints w/ helper classes included in the folder

markstory replied:
I'm fine with them being in serializers, as that directory also contains DRF serializers for request parsing. The top level directory for endpoints was something we tried where parts of the product could be separate django applications, but it didn't really get that much traction.

markstory commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+                "quantity": self.accepted.quantity,
+                "times_seen": self.accepted.times_seen,
+            },
+            "filtered": {"quantity": self.filtered.quantity, "times_seen": self.filtered.quantity},
```
Could you use the datacategory to decide whether to serialize quantity or times_seen as the output `value`? That would make frontend code simpler and reduce your payload size as we don't need both for each category.

JoshFerge replied:
definitely.

markstory commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+
+        # TODO: see if there's a better way to get the user's projects
+        project_list = []
+        team_list = Team.objects.get_for_user(organization=project.organization, user=request.user)
```
You can/should use `OrganizationEndpoint.get_projects()` for this. It handles both kinds of 'all' and organization permissions around open team membership which would be needed here too.

JoshFerge replied:
👍🏼

markstory commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,29 @@
+    )
+
+    for row in result["data"]:
+        row["category"] = (
```
This is likely simpler than trying to cook up a case statement that will yield a similar result in clickhouse.

markstory commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,29 @@
+
+def query(groupby, start, end, rollup, aggregations, orderby, fields=None, filter_keys=None):
+
+    # TODO: if roll up >= use hourly, else use raw
```
You might also want to move the rollup constraint here instead of in the endpoint. That will help with not making invalid queries if this module is used elsewhere in the future.

markstory commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+            rollup=rollup,
+            groupby=["category", "time", "outcome", "reason"],
+            aggregations=[["sum", "times_seen", "times_seen"], ["sum", "quantity", "quantity"]],
+            filter_keys={"org_id": [organization.id]},
```
Should this also filter by the projects you're allowed to see, but not expose that information in the breakdowns?

markstory commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+        return Response(response.build_fields())
+
+
+class OrganizationProjectStatsIndex(OrganizationEndpoint):
```
You might be able to combine this with the Project specific endpoint if the client sends a projectid parameter in the request. The base organization endpoint's `get_projects()` will interpret no project parameter as 'all projects I have membership in'. You can also send -1 from the client to access all projects you have access to. Another benefit of using that approach is that project permissions are enforced more consistently with the rest of the product.

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+        DataCategory.ERROR: (lambda s: s.errors),
+        DataCategory.TRANSACTION: (lambda s: s.transactions),
+        DataCategory.ATTACHMENT: (lambda s: s.attachments),
+    }
```
I take it this is based on the pattern I used in the `UsageStat` class in `billing_usage.py`? It's really only worth it if you want to expose the regular attributes while also offering the `get` method (in my case, I started with the attributes and was forced to bolt on the `get` afterward).

If you can get by with only the `get` (which seems to be the case here at a glance), I'd recommend skipping the attributes entirely and just having an internal dict, keyed by the categories. That simplifies a few things:

```python
    def __init__(self, start, end, rollup):
        self._values = {k: TimeSeriesValues(start, end, rollup) for k in CATEGORY_NAME_MAP}

    def get(self, category):
        return self._values[category]

    def __iter__(self):
        return iter(self._values.items())
```
etc.

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+            self.accepted.times_seen = row["times_seen"]
+        elif row["outcome"] == Outcome.FILTERED:
+            self.filtered.quantity = row["quantity"]
+            self.filtered.times_seen = row["times_seen"]
```
Suggested refactoring. A few more lines but it trims down the repetition.

```python
    def update(self, row):
        def find_measure():
            if row["outcome"] == Outcome.RATE_LIMITED:
                if row["reason"] in {"usage_exceeded", "grace_period"}:
                    return self.over_quota, True
                elif row["reason"] == "smart_rate_limit":
                    return self.spike_protection, False
                else:
                    return self.other, False
            elif row["outcome"] == Outcome.ACCEPTED:
                return self.accepted, True
            elif row["outcome"] == Outcome.FILTERED:
                return self.filtered, True
            else:
                return None, None  # or raise an error?

        measure_to_update, use_times_seen = find_measure()
        if measure_to_update:
            measure_to_update.quantity = row["quantity"]
            measure_to_update.times_seen = row["times_seen" if use_times_seen else "quantity"]
```

JoshFerge replied:
👍🏼 will add that in. going to set quantity and times_seen to the row values here and return a singular value in serialization based on the DataCategory, so helper function will just return the one value.

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+            data_by_time[key] = []
+        else:
+            val = MeasureValue(key)
+            rv[key] = val
```
It seems like `zerofill` is doing a lot of work up front and it's a little hard to follow. Is it only filling in empty entries in case the corresponding timestamps don't later show up in the data? If so, is there a good way to push that to the serialization step, starting instead with empty lists/dicts?

Just a thought. I'll take your word for it if you say that's more trouble than it's worth.

JoshFerge replied:
yeah, how it is now isn't ideal. i think ideally we move it into the outcomes snuba query, and this code can gracefully create an object and this complexity can be removed from the class.

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+            if row["reason"] in {"usage_exceeded", "grace_period"}:
+                self.over_quota.quantity = row["quantity"]
+                self.over_quota.times_seen = row["times_seen"]
+            elif row["reason"] in "smart_rate_limit":
```
Should be `row["reason"] == "smart_rate_limit"`?

JoshFerge replied:
yup 👍🏼

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,239 @@
+        self.filtered = StatMeasure(0, 0)
+        self.over_quota = StatMeasure(0, 0)
+        self.spike_protection = StatMeasure(0, 0)
+        self.other = StatMeasure(0, 0)
```
Nit: I'd write it as `StatMeasure()`.

JoshFerge replied:
👍🏼

RyanSkonnord reviewed (commented):
👍 on the OO refactor. I hope you agree that it's a bit easier to work with this way.

Names are hard. 😉

JoshFerge committed:
PR feedback

JoshFerge committed:
rename classes

JoshFerge committed:
name improvements

JoshFerge committed:
rollup ergonomics, combinine proj endpoints

JoshFerge committed:
support less than hour queries"
;

JoshFerge committed:
Merge branch 'master' into joshferge/outcomes-snuba-module

JoshFerge committed:
remove times_seen field, move around rollup validation

JoshFerge committed:
make orgstats api more flexible

add bad interval test

make outcomes api like sessions api

add invalid tests

all tests passing

more tests

get conditions working

add parameter validation

fix tests

JoshFerge committed:
cleanup, return 'spike_protection' to client

JoshFerge committed:
Merge branch 'master' into joshferge/outcomes-snuba-module

JoshFerge committed:
code cleanups, add sub hr queries

JoshFerge committed:
small refactors

JoshFerge committed:
more code cleanup

JoshFerge committed:
add spans, referrer

JoshFerge committed:
add typing

JoshFerge committed:
small refactor

JoshFerge committed:
add more tests

JoshFerge committed:
remove key_id logic

JoshFerge committed:
fix remaining typing issues

JoshFerge committed:
small edits, fix broken tests

JoshFerge committed:
fix tsdb tests

JoshFerge committed:
fix more tests

JoshFerge committed:
test improvements

JoshFerge committed:
code touchups

JoshFerge committed:
merge with master

JoshFerge commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,337 @@
+        return ["quantity"]
+
+    def extract_from_row(
+        self, row: Dict[str, Union[Any, int]], group: Optional[Dict[str, Any]] = None
```
have to do this weird Union to get mypi to like having this function returning an int. wonder what the best way to to this is.

JoshFerge replied:
fixed, went ahead and just typecasting the return value to `int`, which is essentially a noop here. h/t @RyanSkonnord

JoshFerge committed:
fix tests after master merge

JoshFerge committed:
fix typing of extract_from_row

JoshFerge commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,67 @@
+        )
+
+    @contextmanager
+    def handle_query_errors(self):
```
this bit of TODO was copied over from the organization_sessions endpoint. its probably worth it to refactor and introduce a new endpoint base that sessions and this one can use for these inherited methods, but didn't want this PR to balloon, so keeping it as is for now.

JoshFerge commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,330 @@
+from .dataset import Dataset
+from sentry_relay import DataCategory
+
+from sentry.snuba.sessions_v2 import (
```
there is a potential to move some of this (and other funcs within other snuba modules) into a common that the endpoints can import / inherit from, but abstracting out didn't feel worth it here, so just importing from sessions directly.

evanh commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,330 @@
+        raise NotImplementedError()
+
+    @abstractmethod
+    def map_row(self, row: Dict[str, Any]) -> None:
```
Here, since `row` is modified by the function `Dict` or `MutableMapping` works.

JoshFerge replied:
👍🏼

evanh commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,330 @@
+
+def get_filter(
+    query: QueryDict, params: Dict[Any, Any]
+) -> Tuple[List[Tuple[str, str, List[Any]]], Dict[str, Any]]:
```
This is where a NamedTuple for a single Condition would clean this up a bit. Or if returning conditions and filter keys in a tuple happens a lot, wrapping that in a NamedTuple.

JoshFerge replied:
Tried to make it a namedtuple, but looks like our snuba json serializer likes to convert namedtuples to dicts, which made snuba unhappy about that. Figured it's best not to jump down that rabbit hole so just using a regular tuple type variable, definitely looks cleaner. thanks!

evanh commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,330 @@
+        self.conditions, self.filter_keys = get_filter(query, params)
+
+
+def run_outcomes_query(query: QueryDefinition) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
```
Adding `ResultSet = List[Dict[str, Any]]` would clean up a lot of these functions.

JoshFerge replied:
100%. updated! looks much better now.

evanh reviewed (commented):
Some typing suggestions. Way to be a pioneer!

JoshFerge committed:
typing improvements

JoshFerge commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,330 @@
+
+class Field(ABC):
+    @abstractmethod
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
```
got it! makes sense, thanks.

JoshFerge commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,330 @@
+        return ["quantity"]
+
+    def extract_from_row(self, row: Dict[str, Any], group: Optional[Dict[str, Any]] = None) -> int:
+        if row is None:
```
aha this one was not quite right. added an optional wrapper for that. thanks for the catch!

evanh replied:
If you run mypy with the `--warn-unreachable` flag I think it should have surfaced this as well.

JoshFerge commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,339 @@
+def _format_rows(rows: ResultSet, query: QueryDefinition) -> ResultSet:
+    category_grouping: Dict[str, Any] = {}
+
+    def _group_row(row: Dict[str, Any]) -> None:
```
will definitely want to see performance of this in staging. there are some easier optimizations to be had with this function (we only need to do it on rows category = error, security, default). would be an issue in orgs with large project cardinality. additionally, it could be done in the massaging step so we're not looping through twice.

JoshFerge replied:
there is also the option of doing the groupby in the snuba query instead of here, which is also an option.

JoshFerge committed:
add hour const

RyanSkonnord commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,67 @@
+    def handle_query_errors(self):
+        try:
+            # TODO: this context manager should be decoupled from `OrganizationEventsEndpointBase`?
+            with super().handle_query_errors():
```
It looks like `OrganizationEventsEndpointBase.handle_query_errors` is logically a static method (that is, its body does not refer to `self`). 👍 both to pulling it into a new base class and to deferring it outside this PR. But, if you're looking for another way, it might be fine as a top-level function if you want to avoid inheritance entirely.

JoshFerge replied:
will add this to my few pieces of tech debt bill from this PR, but making a static method seems 👌🏼

RyanSkonnord commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,341 @@
+            return ["sum", "times_seen", "times_seen"]
+        else:
+            # RawOutcomes doesnt have times_seen, do a count instead
+            return ["count()", "", "times_seen"]
```
What's the deal with the three aggregation values? Do the three positions have any significance? Is it documented anywhere?

A dataclass with three fields could improve on `List[str]`, but probably not worth the extra lines without 3.8's dataclass feature.

If the value is constrained to always have the same length, I think a tuple matches Python conventions better than a list. You can do `Tuple[str, str, str]` as the type hint, right?

evanh replied:
Just FYI, dataclasses are available as a third party library in Python <3.8.

> You can do Tuple[str, str, str] as the type hint, right?

Just a note that since the return value isn't a Tuple mypy wouldn't like that type hinting.

JoshFerge replied:
@RyanSkonnord that piece is documented in `SnubaQueryParams` in `utils/snuba.py` : 
> `aggregations` a list of (aggregation_function, column, alias) tuples to be passed to the query.

Definitely agree this could be more obvious with types / NamedTuples / DataClasses. Tried to use some NamedTuples for some of the types but we JSON serialize NamedTuples as JSON objects instead of an array, so didn't work out there. For now, i've converted this return type to Tuple[str,str,str] as it is more descriptive.

RyanSkonnord commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,341 @@
+            if DataCategory.parse(category) == DataCategory.ERROR:
+                resolved_categories.update(DataCategory.error_categories())
+            else:
+                resolved_categories.add(DataCategory.parse(category))
```
I think I've seen this pattern around elsewhere. Any chance a `DataCategory` class method like the following could be used from more than one place?
```python
def parse_categories(cls, category):
    dc = cls.parse(category)
    return cls.error_categories() if dc == cls.ERROR else [dc]
```

JoshFerge replied:
agreed it could be useful elsewhere, but yes also agree out of scope of this changeset.

JoshFerge committed:
better field validation

JoshFerge commented on src/sentry/api/endpoints/organization_stats_v2.py:
```diff
@@ -0,0 +1,67 @@
+            return Response(result, status=200)
+
+    def build_outcomes_query(self, request, organization):
+        for param in request.GET:
+            if param not in VALID_PARAMETERS:
+                raise InvalidParams(f'Invalid parameter: "{param}"')
```
👍🏼 will go ahead and remove this bit. and the various parameters end up being validated by the helper functions.

JoshFerge commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,341 @@
+"""
+The new Outcomes API defines a "metrics"-like interface which is can be used in
+a similar way to "sessions" and "discover"
+# "Metrics"
```
👍🏼

JoshFerge committed:
pr feedback

JoshFerge commented on src/sentry/snuba/outcomes.py:
```diff
@@ -0,0 +1,341 @@
+        conditions=query.conditions,
+        selected_columns=query.query_columns,
+        referrer="outcomes.totals",
+        # orderby=query.orderby, TODO: add orderby?
```
we end up sorting in the post processing function, will keep an eye on this when testing performance

JoshFerge committed:
pr feedback

JoshFerge committed:
Merge branch 'master' into joshferge/outcomes-snuba-module

JoshFerge committed:
fix outcomes test

JoshFerge committed:
Merge branch 'master' into joshferge/outcomes-snuba-module

RyanSkonnord reviewed (approved):
Implementation looks good to me. Someone else might still take a pass to check feature/integration correctness.

JoshFerge committed:
Merge branch 'master' into joshferge/outcomes-snuba-module

JoshFerge committed:
fix conflict

JoshFerge committed:
Merge branch 'master' into joshferge/outcomes-snuba-module

JoshFerge committed:
fix span name, correct error import
